File: recursion/c001.java
import java.util.*;
class c001{
    public static void main(String[] args){
        String s="abc";
        subsequence_void(s,"");
        //System.out.println(ans);
    }

    public static ArrayList<String> subsequence_ret(String s){
        if(s.length()==0){
            ArrayList<String> ans=new ArrayList<String>();
            ans.add("");
            return ans;
        }

        char ch=s.charAt(0);
        ArrayList<String> smallans=subsequence_ret(s.substring(1));
        ArrayList<String> ans=new ArrayList<String>();
        for(String ns:smallans){
            ans.add(ns);
            ans.add(ch+ns);
        }      
        return ans; 
    }
    public static void subsequence_void(String str,String ans){
        if(str.length()==0){
            System.out.println(ans);
            return;
        }

        //int count=0;
        subsequence_void(str.substring(1),ans);
        char ch=str.charAt(0);
        subsequence_void(str.substring(1), ans+ch);
        return;
    }
}

File: recursion/c001.cpp
#include<iostream>
#include<vector>
using namespace std;


vector<string> subsequence(string s){
    vector<string> ans;
    if(s.size()==0){
        ans.push_back("");
        return ans;
    }
    char ch=s[0];
    vector<string> smallans=subsequence(s.substr(1));
    for(string sa:smallans){
        ans.push_back(sa);
        ans.push_back(ch+sa);
    }
    return ans;
}

int main(){
    string s = {"abc"};
    vector<string> ans= subsequence(s);
    for(int i=0;i<ans.size();i++){
        cout<<ans[i]<<endl;
    }
    return 0;
}